app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshake.kt

```diff
@@ -0,0 +1,151 @@
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver. Finally, the onPostExecute() of the sender
+ * initiates the file transfer through [SenderDevice] on the sender and using
+ * [ReceiverDevice] on the receiver.
+ */
+internal class PeerGroupHandshake(private val wifiDirectManager: WifiDirectManager) {
+  private val HANDSHAKE_MESSAGE = "Request Kiwix File Sharing"
+  suspend fun peer(): InetAddress? = withContext(Dispatchers.IO) {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Handshake in progress")
+    }
+    if (wifiDirectManager.isGroupFormed && wifiDirectManager.isGroupOwner && this.isActive) {
+      try {
+        ServerSocket(PEER_HANDSHAKE_PORT)
+          .use { serverSocket ->
+            serverSocket.reuseAddress = true
+            val server = serverSocket.accept()
+            val objectInputStream = ObjectInputStream(server.getInputStream())
+            val kiwixHandShakeMessage = objectInputStream.readObject()
+
+            // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+            return@withContext if (isKiwixHandshake(kiwixHandShakeMessage) && this.isActive) {
+              if (BuildConfig.DEBUG) {
+                Log.d(TAG, "Client IP address: " + server.inetAddress)
+              }
+              exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream())
+              server.inetAddress
+            } else {
+              // Selected device is not accepting wifi direct connections through the kiwix app
+              null
+            }
+          }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return@withContext null
+      }
+    } else if (wifiDirectManager.isGroupFormed && this.isActive) { // && !groupInfo.isGroupOwner
+      try {
+        Socket().use { client ->
+          client.reuseAddress = true
+          client.connect(
+            InetSocketAddress(
+              wifiDirectManager.groupOwnerAddress.hostAddress,
+              PEER_HANDSHAKE_PORT
+            ), 15000
+          )
+          val objectOutputStream = ObjectOutputStream(client.getOutputStream())
+          // Send message for the peer device to verify
+          objectOutputStream.writeObject(HANDSHAKE_MESSAGE)
+          exchangeFileTransferMetadata(client.getOutputStream(), client.getInputStream())
+          return@withContext wifiDirectManager.groupOwnerAddress
+        }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return@withContext null
+      }
+    }
+    return@withContext null
+  }
+
+  private fun isKiwixHandshake(handshakeMessage: Any): Boolean =
+    HANDSHAKE_MESSAGE == handshakeMessage
+
+  private fun exchangeFileTransferMetadata(outputStream: OutputStream, inputStream: InputStream) {
```

Comment:
I also wouldn't object to this class becoming abstract and instead of this method relying on an if we have ReceiverHandshakeTak/SenderHandshakeTask .

`wifiDirectManager.isGroupOwner` is just an analog for `wifiDirectManager.isFileSender` I am pretty sure because the sender always initiates, forming the group, so the above method can be split among the subclasses too.

I am not sure when all is said and done there will be a need for a base class at all, 2 very separate things happen here. Maybe it will be nice on the client side eg `val task:PeerGroupHandshakeTask = if(isSender) SenderHandshake() else ReceiverHandshake()`, probably better have them conform to an interface though

Reply:
I am not sure I'm getting 100% of what you are saying in here. But I tried in [here](https://github.com/kiwix/kiwix-android/pull/2389/commits/6853e82b76a28a73e7b96b96df7c0c044386b148)

Reply:
I was in my first paragraph talking about making an abstract class `PeerGroupHandshake` with an abstract method `exchangeFileTransferMetadata` and 2 instance of this class `ReceiverHandshakeTask` `SenderHandshakeTask` which contain the separate implementations.

In my second paragraph I am taking a guess that

```
if (wifiDirectManager.isGroupFormed && wifiDirectManager.isGroupOwner && this.isActive) {
    //owner code block
    } else if (wifiDirectManager.isGroupFormed && this.isActive) {
//client code block
}
```
isSender/Receiver has an analogy with owner/client here so additionally we could make `handshake` an abstract method. I am not sure if this is true but you may be able to find out via experimentation or reading on wifi group formation.


in my 3rd paragraph I realise that if the 2nd paragraph is the case then these classes have no business sharing an abstract class because they share no logic or implementation details, at best 1 constant so `PeerGroupHandshakeTask` can be converted to an interface with zero issue, well you probably have to pass the wifiDirectManager as an argument to the method `handshake` instead of a constructor param.

Does that clear things up or do you want more concrete examples?

Reply:
>do you want more concrete examples?

that would be great!

Reply:
okay so lets give an example somewhat along the lines of my first paragraph.

```
abstract class AnimalTask(){

fun play(){
  getPetted()
  makeSound()
}

fun getPetted(){
 println("hooray pets")
}

abstract fun makeSound()
```

```
class DogTask: AnimalTask(){
  override fun makeSound(){
    println("woof")
  }
}
```

```
class CatTask: AnimalTask(){
  override fun makeSound(){
    println("meow")
  }
}
```

Here I have introduced an abstract class because functionality is shared between both animal tasks and the subclasses only implement the functionality that differs between them. It could be used like so

```
val animalTask = if (isDog) DogTask() else CatTask()
animalTask.play()
```

Now moving on, this code looks a little fishy to me!
```
fun getPetted(){
 println("hooray pets")
}
```
That doesn't sound like the inner monologue of a cat I've ever met so lets change it

```
abstract class AnimalTask(){

fun play(){
  getPetted()
  makeSound()
}

abstract fun getPetted()

abstract fun makeSound()

}

class DogTask: AnimalTask(){
override fun getPetted(){
 println("hooray pets")
}

  override fun makeSound(){
    println("woof")
  }
}

class CatTask: AnimalTask(){

  override fun getPetted(){
 println("how dare you")
}

  override fun makeSound(){
    println("meow")
  }
}
```

Now it seems like these 2 separate tasks don't have very much in common at all, aside from they get petted and then make a sound, if we tried to model this further this might not even hold true. The dog might roll over and the cat could just ignore you, all we really want to inforce on them is we tried to play with them so we can now define it as

```
interface AnimalTask{
fun play()
}

class DogTask: AnimalTask{
override fun play(){
  println("hooray pets")
  println("roll over now")
  println("woof")
}

class CatTask: AnimalTask{
override fun play(){
  println("how dare you")
  println("haughtily ignore")
  println("hiss")
}
```

Now aside from how much I dislike cats, have you learned anything new?

Reply:
Thanks for the advice. I try to implement as you said. By the way nice example 😄 .

Reply:
glad to be of help